plugins/gateway/tyk/v3.2.2/plugin.go

```diff
@@ -0,0 +1,190 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"unsafe"
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/TykTechnologies/tyk/ctx"
+	"github.com/TykTechnologies/tyk/log"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	uuid "github.com/satori/go.uuid"
+
+	"github.com/apiclarity/apiclarity/plugins/api/client/client"
+	"github.com/apiclarity/apiclarity/plugins/api/client/client/operations"
+	"github.com/apiclarity/apiclarity/plugins/api/client/models"
+)
+
+var logger = log.Get()
+
+// Called during post phase for setting the apiDefinition since we dont get it in the response phase.
+func PostGetAPIDefinition(rw http.ResponseWriter, r *http.Request) {
+	apiDefinition := ctx.GetDefinition(r)
+	if apiDefinition == nil {
+		apiDefinition = apiDefinitionRetriever(r.Context())
+	}
+
+	if apiDefinition == nil {
+		logger.Error("Failed to get api definition")
+		return
+	}
+	ctx.SetDefinition(r, apiDefinition)
+}
+
+// Called during response phase
+func ResponseSendTelemetry(rw http.ResponseWriter, res *http.Response, req *http.Request) {
+	logger.Info("Handling telemetry")
+	telemetry, err := createTelemetry(res, req)
+	if err != nil {
+		logger.Errorf("Failed to create telemetry: %v", err)
+		return
+	}
+
+	apiClient := newAPIClient("apiclarity.apiclarity:9000")
+	params := operations.NewPostTelemetryParams().WithBody(telemetry)
+
+	_, err = apiClient.Operations.PostTelemetry(params)
+	if err != nil {
+		logger.Errorf("Failed to post telemetry: %v", err)
+		return
+	}
+	B, _ := json.Marshal(telemetry)
+	logger.Infof("Telemetry has been sent: %s", B)
+}
+
+func createTelemetry(res *http.Response, req *http.Request) (*models.Telemetry, error) {
+	apiDefinition := ctx.GetDefinition(req)
+	if apiDefinition == nil {
+		return nil, fmt.Errorf("failed to get api definition in response")
+	}
+
+	host, port := getHostAndPortFromTargetURL(apiDefinition.Proxy.TargetURL)
+
+	reqBody, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body: %v", err)
+	}
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %v", err)
+	}
+
+	telemetry := models.Telemetry{
+		DestinationAddress:   ":" + port, // No destination ip for now
+		DestinationNamespace: "",
+		Request: &models.Request{
+			Common: &models.Common{
+				TruncatedBody: false,
+				Body:          strfmt.Base64(reqBody),
+				Headers:       createHeaders(req.Header),
+				Version:       req.Proto,
+			},
+			Host:   host,
+			Method: req.Method,
+			Path:   req.URL.String(),
```

Comment:
DYT it is safer to create it with `req.URL.Path + "?" + req.URL.RawQuery` in case `req.URL.RawQuery` is not empty.
`req.URL.String()` can include also `Scheme`, `User` and `Fragment`